ethcore/src/client/client.rs

```diff
@@ -363,23 +353,17 @@ impl Client {
 
 			for block in blocks {
 				let header = &block.header;
-				if invalid_blocks.contains(header.parent_hash()) {
-					invalid_blocks.insert(header.hash());
-					continue;
-				}
-				let closed_block = self.check_and_close_block(&block);
-				if let Err(_) = closed_block {
-					invalid_blocks.insert(header.hash());
-					continue;
-				}
+				let is_invalid = invalid_blocks.contains(header.parent_hash());
+				if let (false, Ok(closed_block)) = (is_invalid, self.check_and_close_block(&block)) {
```

Comment:
`check_and_close_block` must not be called if `invalid_blocks.contains(header.parent_hash())`